packages/shared/src/error.ts

```diff
@@ -0,0 +1,63 @@
+const ERROR_CODES = {
+  INTERNAL_SERVER_ERROR: 500,
+  INTERNAL_CLIENT_ERROR: 500,
+  FAILED_TO_SIGN: 500,
+  FAILED_TO_UPLOAD: 500,
+  BAD_REQUEST: 400,
+  MISSING_ENV: 500,
+  FILE_LIMIT_EXCEEDED: 500,
+};
+
+type ErrorCode = keyof typeof ERROR_CODES;
+
+function messageFromUnknown(cause: unknown) {
+  if (typeof cause === "string") {
+    return cause;
+  }
+  if (cause instanceof Error) {
+    return cause.message;
+  }
+  if (cause && typeof cause === "object" && "message" in cause) {
+    return cause.message;
+  }
+}
+
+export class UploadThingError extends Error {
+  public readonly cause?: Error;
+  public readonly code: ErrorCode;
+
+  constructor(opts: {
+    code: keyof typeof ERROR_CODES;
+    message?: string;
+    cause?: unknown;
+  }) {
+    const message = opts.message ?? messageFromUnknown(opts.cause) ?? opts.code;
+
+    // @ts-expect-error https://github.com/tc39/proposal-error-cause
+    super(message, { cause: opts.cause });
+
+    this.code = opts.code;
+  }
+
+  public static async fromResponse(response: Response) {
+    const message = await response.text();
+    return new UploadThingError({
+      message,
+      code: getErrorCodeFromStatusCode(response.status),
+      cause: response,
```

Comment:
I am pretty sure `messageFromUnknown()` can't resolve this?

Reply:
There are a few places we are passing a fetch response as a cause, we should probably handle it?

Reply:
> I am pretty sure messageFromUnknown() can't resolve this?
Since we're passing a message we don't need to worry about messageFromUnknown

> There are a few places we are passing a fetch response as a cause, we should probably handle it?
Yes - I have a feeling a type is wrong which is leaking unknown somewhere

Reply:
> Since we're passing a message we don't need to worry about messageFromUnknown

Right 🤦